File: 2023/01/main.c
#include <stdio.h>
#include <string.h>
#include <ctype.h>
#include <assert.h>

#define ARRAY_SIZE(XS) (sizeof(XS) / sizeof((XS)[0]))

struct number {
    const char* cstr;
    size_t len;
    size_t value;
};

static const struct number numbers[] = {
    { .cstr = "one",   .len = sizeof("one")   - 1, .value = 1 },
    { .cstr = "two",   .len = sizeof("two")   - 1, .value = 2 },
    { .cstr = "three", .len = sizeof("three") - 1, .value = 3 },
    { .cstr = "four",  .len = sizeof("four")  - 1, .value = 4 },
    { .cstr = "five",  .len = sizeof("five")  - 1, .value = 5 },
    { .cstr = "six",   .len = sizeof("six")   - 1, .value = 6 },
    { .cstr = "seven", .len = sizeof("seven") - 1, .value = 7 },
    { .cstr = "eight", .len = sizeof("eight") - 1, .value = 8 },
    { .cstr = "nine",  .len = sizeof("nine")  - 1, .value = 9 },
};

size_t calibration_value(const char* line, size_t line_len);
size_t ascii_char_to_num(char c);

int main(void)
{
    size_t total_calibration_value = 0;

    char line[512];
    while (fgets(line, sizeof(line), stdin) != NULL) {
        size_t line_len = strnlen(line, sizeof(line));
        assert(line_len > 0);
        line[--line_len] = '\0';
        
        total_calibration_value += calibration_value(line, line_len);
    }

    printf("sum: %zu\n", total_calibration_value);
}

size_t calibration_value(const char* line, size_t line_len)
{
    size_t result = 0;

    for (const char* start = line; *start != '\0'; start++) {
        if (isdigit(*start)) {
            result += 10 * ascii_char_to_num(*start);
            break;
        }
        for (size_t i = 0; i < ARRAY_SIZE(numbers); i++) {
            if (strncmp(start, numbers[i].cstr, numbers[i].len) == 0) {
                result += 10 * numbers[i].value;
                goto scan_start_done;
            }
        }
    }
scan_start_done:

    for (const char* end = line + line_len - 1; end >= line; end--) {
        if (isdigit(*end)) {
            result += ascii_char_to_num(*end);
            break;
        }
        for (size_t i = 0; i < ARRAY_SIZE(numbers); i++) {
            const char* start = end - numbers[i].len + 1;
            if (start < line) {
                continue;
            }
            if (strncmp(start, numbers[i].cstr, numbers[i].len) == 0) {
                result += numbers[i].value;
                goto scan_end_done;
            }
        }
    }
scan_end_done:

    return result;
}

size_t ascii_char_to_num(char c)
{
    assert(c >= '0' && c <= '9');
    return (size_t) c - '0';
}

File: 2023/01/main.ml
open Printf
open Unix

(* Utilities *)

let stdin_size: int = (Unix.fstat Unix.stdin).st_size

let stdin_read_all () : string = 
    let ch = open_in_bin "/dev/stdin" in
    let finally () = close_in ch in
    Fun.protect ~finally (fun () ->
        really_input_string ch stdin_size
    )

let is_not_blank (s: string) : bool =
    let trimmed = String.trim s in
    String.length trimmed > 0

let is_digit = function
    | '0' .. '9' -> true
    | _ -> false

let list_of_string (s: string) : char list =
    List.init (String.length s) (String.get s)

let seq_fold_sum = Seq.fold_left (+) 0

(* Solutions *)

let calibration_value_a (s: string) : int =
    let first: char = match s |> String.to_seq |> Seq.find is_digit with
        | None -> failwith "line is expected to contain at least a number"
        | Some c -> c
    in
    let last: char = list_of_string s
        |> List.rev
        |> List.find is_digit
    in
    int_of_string (sprintf "%c%c" first last)

let literals : string list = [
    "one"  ;
    "two"  ;
    "three";
    "four" ;
    "five" ;
    "six"  ;
    "seven";
    "eight";
    "nine" 
]

let literals_table : (string, char) Hashtbl.t =
    let table = Hashtbl.create 9 in
    let value = ref 0 in
    literals |> List.iter (fun lit ->
        incr value;
        let ascii = !value + (Char.ord '0') in
        Hashtbl.add table lit ascii
    );
    table

let calibration_value_b (s: string) : int =
    let first_char = String.get s 0 in
    (*
    - check if first char is a digit. if it is, then it's the char
    - iterate over the literals keys and check if
    *)
    let digit = if is_digit first_char then first_char else
        let rec aux () =
        in
        aux 
    in

let solve_a (lines: string list) : int =
    lines
    |> List.to_seq
    |> Seq.filter is_not_blank
    |> Seq.map calibration_value_a
    |> seq_fold_sum

let solve_b (lines: string list) : int =
    lines
    |> List.to_seq
    |> Seq.filter is_not_blank
    |> Seq.map calibration_value_b
    |> seq_fold_sum

let () =
    (* input reading and parsing *)
    let input: string = stdin_read_all () in
    let lines: string list = String.split_on_char '\n' input in

    (* Part A *)
    let a_solution: int = solve_a lines in
    printf "Part A: %d\n" a_solution;

    (* Part B *)
    let b_solution: int = solve_b lines in
    printf "Part B: %d\n" b_solution

